src/vdaf.rs

```diff
@@ -166,9 +166,10 @@ pub trait Vdaf: Clone + Debug {
 pub trait Client: Vdaf {
     /// Shards a measurement into a public share and a sequence of input shares, one for each
     /// Aggregator.
-    fn shard(
+    fn shard<const N: usize>(
```

Comment:
If we put the generic separately on `Client::shard` and `Aggregator::prepare_init`, then there's no enforcement that the same `N` is used. Instead, I think we should have `pub trait Vdaf<const N: usize>` and then `pub trait Client<const N: usize>: Vdaf<N>` and `pub trait Aggregator<const L: usize, const N: usize>: Vdaf<N>` and then all the other places the compiler complains about.

Reply:
Yeah I think you're right. That approach seems general enough for any solution for https://github.com/cfrg/draft-irtf-cfrg-vdaf/issues/127. I'm gonna send a PR to address that issue, then will update here.

Reply:
As discussed in https://github.com/divviup/libprio-rs/pull/455#discussion_r1113368715, I've added a new const generic for the nonce size. I decided to put this on Client and Aggregator rather than VDAF. My thinking is that the Collector won't ever use the nonce length, so it doesn't make sense to add this as a trait bound.

On the other hand, it might be useful to put both `NONCE_SIZE` and `VERIFY_KEY_SIZE` on Vdaf so that the API looks a bit more like the spec. My only objection would be that it *seems to me* to be bad form to add bounds on traits that aren't strictly needed.

Reply:
The downside of putting the generic on `trait Client` and `Aggregator` is that now I can do this (very rough pseudocode):

```
let my_client: Client<NONCE_SIZE = 8> = whatever_mod::new_client();
let my_aggregator: Aggregator<NONCE_SIZE = 16> = whatever_mod::new_aggregator();

let (pub, shares) = my_client.shard(...)?;
let _ = my_aggregator.prep_init(...)?;
```

This won't work, because despite implementing the same `Vdaf`, the client and aggregator use different nonce sizes. I won't find out that it doesn't work until runtime. Now, if `NONCE_SIZE` were on `trait Vdaf`, then the compiler can statically verify that `my_client` and `my_aggregator` are using the same nonce size.

Now, this is a bit of a reach, because this scenario only comes up when writing tests, because otherwise VDAF clients and aggregators are going to be compiled into separate executables and the compiler can't enforce consistency in that scenario. So I'll defer to your preference as the author here.

Reply:
I see how this could happen if the implementation of the `Client` and `Aggregator` were generic in the nonce size. However I imagine this will usually be hard-coded. (For example, in this PR, `Prio3` implements `Client<16>`, not `Client<NONCE_SIZE>`. We don't need this to be generic because the VDAF spec spells out the nonce size; any other choice of nonce size would not be compliant.) 

I tacked on a commit to see what this would look like. Overall the change isn't too bad. I would say the only major downside is that `fieldvec_roundtrip_test` now needs `VERIFY_KEY_SIZE` and `NONCE_SIZE` as parameters. To me this makes the code harder to read.

Reply:
I've reverted the commit with this change. For the record, here is the patch:
```
diff --git a/src/vdaf.rs b/src/vdaf.rs
index 5c3beac..c95a39f 100644
--- a/src/vdaf.rs
+++ b/src/vdaf.rs
@@ -126,7 +126,7 @@ impl<F: FieldElement, const L: usize> Encode for Share<F, L> {
 /// The base trait for VDAF schemes. This trait is inherited by traits [`Client`], [`Aggregator`],
 /// and [`Collector`], which define the roles of the various parties involved in the execution of
 /// the VDAF.
-pub trait Vdaf: Clone + Debug {
+pub trait Vdaf<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>: Clone + Debug {
     /// Algorithm identifier for this VDAF.
     const ID: u32;

@@ -163,7 +163,9 @@ pub trait Vdaf: Clone + Debug {
 }

 /// The Client's role in the execution of a VDAF.
-pub trait Client<const NONCE_SIZE: usize>: Vdaf {
+pub trait Client<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>:
+    Vdaf<VERIFY_KEY_SIZE, NONCE_SIZE>
+{
     /// Shards a measurement into a public share and a sequence of input shares, one for each
     /// Aggregator.
     fn shard(
@@ -174,7 +176,9 @@ pub trait Client<const NONCE_SIZE: usize>: Vdaf {
 }

 /// The Aggregator's role in the execution of a VDAF.
-pub trait Aggregator<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>: Vdaf {
+pub trait Aggregator<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>:
+    Vdaf<VERIFY_KEY_SIZE, NONCE_SIZE>
+{
     /// State of the Aggregator during the Prepare process.
     type PrepareState: Clone + Debug;

@@ -231,7 +235,9 @@ pub trait Aggregator<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>: Vda
 }

 /// The Collector's role in the execution of a VDAF.
-pub trait Collector: Vdaf {
+pub trait Collector<const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>:
+    Vdaf<VERIFY_KEY_SIZE, NONCE_SIZE>
+{
     /// Combines aggregate shares into the aggregate result.
     fn unshard<M: IntoIterator<Item = Self::AggregateShare>>(
         &self,
@@ -341,7 +347,7 @@ pub(crate) fn run_vdaf<V, M, const L: usize>(
     measurements: M,
 ) -> Result<V::AggregateResult, VdafError>
 where
-    V: Client<16> + Aggregator<L, 16> + Collector,
+    V: Client<L, 16> + Aggregator<L, 16> + Collector<L, 16>,
     M: IntoIterator<Item = V::Measurement>,
 {
     use rand::prelude::*;
@@ -407,7 +413,7 @@ pub(crate) fn run_vdaf_prepare<V, M, const L: usize>(
     input_shares: M,
 ) -> Result<Vec<V::OutputShare>, VdafError>
 where
-    V: Client<16> + Aggregator<L, 16> + Collector,
+    V: Client<L, 16> + Aggregator<L, 16> + Collector<L, 16>,
     M: IntoIterator<Item = V::InputShare>,
 {
     let input_shares = input_shares
@@ -476,10 +482,13 @@ where
 }

 #[cfg(test)]
-fn fieldvec_roundtrip_test<F, V, T>(vdaf: &V, agg_param: &V::AggregationParam, length: usize)
-where
+fn fieldvec_roundtrip_test<F, V, T, const VERIFY_KEY_SIZE: usize, const NONCE_SIZE: usize>(
+    vdaf: &V,
+    agg_param: &V::AggregationParam,
+    length: usize,
+) where
     F: FieldElement,
-    V: Vdaf,
+    V: Vdaf<VERIFY_KEY_SIZE, NONCE_SIZE>,
     T: Encode,
     for<'a> T: ParameterizedDecode<(&'a V, &'a V::AggregationParam)>,
 {
diff --git a/src/vdaf/poplar1.rs b/src/vdaf/poplar1.rs
index d631f76..9115dd7 100644
--- a/src/vdaf/poplar1.rs
+++ b/src/vdaf/poplar1.rs
@@ -298,7 +298,7 @@ impl Decode for Poplar1AggregationParam {
     }
 }

-impl<P: Prg<L>, const L: usize> Vdaf for Poplar1<P, L> {
+impl<P: Prg<L>, const L: usize> Vdaf<L, 16> for Poplar1<P, L> {
     const ID: u32 = 0x00001000;
     type Measurement = IdpfInput;
     type AggregateResult = Vec<u64>;
@@ -313,7 +313,7 @@ impl<P: Prg<L>, const L: usize> Vdaf for Poplar1<P, L> {
     }
 }

-impl<P: Prg<L>, const L: usize> Client<16> for Poplar1<P, L> {
+impl<P: Prg<L>, const L: usize> Client<L, 16> for Poplar1<P, L> {
     fn shard(
         &self,
         input: &IdpfInput,
@@ -600,7 +600,7 @@ impl<P: Prg<L>, const L: usize> Aggregator<L, 16> for Poplar1<P, L> {
     }
 }

-impl<P: Prg<L>, const L: usize> Collector for Poplar1<P, L> {
+impl<P: Prg<L>, const L: usize> Collector<L, 16> for Poplar1<P, L> {
     fn unshard<M: IntoIterator<Item = Poplar1FieldVec>>(
         &self,
         agg_param: &Poplar1AggregationParam,
diff --git a/src/vdaf/prio2.rs b/src/vdaf/prio2.rs
index ade2ed7..53e522d 100644
--- a/src/vdaf/prio2.rs
+++ b/src/vdaf/prio2.rs
@@ -91,7 +91,7 @@ impl Prio2 {
     }
 }

-impl Vdaf for Prio2 {
+impl Vdaf<32, 16> for Prio2 {
     const ID: u32 = 0xFFFF0000;
     type Measurement = Vec<u32>;
     type AggregateResult = Vec<u32>;
@@ -107,7 +107,7 @@ impl Vdaf for Prio2 {
     }
 }

-impl Client<16> for Prio2 {
+impl Client<32, 16> for Prio2 {
     fn shard(
         &self,
         measurement: &Vec<u32>,
@@ -269,7 +269,7 @@ impl Aggregator<32, 16> for Prio2 {
     }
 }

-impl Collector for Prio2 {
+impl Collector<32, 16> for Prio2 {
     fn unshard<M: IntoIterator<Item = AggregateShare<FieldPrio2>>>(
         &self,
         _agg_param: &Self::AggregationParam,
@@ -463,12 +463,20 @@ mod tests {
     #[test]
     fn roundtrip_output_share() {
         let vdaf = Prio2::new(31).unwrap();
-        fieldvec_roundtrip_test::<FieldPrio2, Prio2, OutputShare<FieldPrio2>>(&vdaf, &(), 31);
+        fieldvec_roundtrip_test::<FieldPrio2, Prio2, OutputShare<FieldPrio2>, 32, 16>(
+            &vdaf,
+            &(),
+            31,
+        );
     }

     #[test]
     fn roundtrip_aggregate_share() {
         let vdaf = Prio2::new(31).unwrap();
-        fieldvec_roundtrip_test::<FieldPrio2, Prio2, AggregateShare<FieldPrio2>>(&vdaf, &(), 31);
+        fieldvec_roundtrip_test::<FieldPrio2, Prio2, AggregateShare<FieldPrio2>, 32, 16>(
+            &vdaf,
+            &(),
+            31,
+        );
     }
 }
diff --git a/src/vdaf/prio3.rs b/src/vdaf/prio3.rs
index 42ebac7..eb109a0 100644
--- a/src/vdaf/prio3.rs
+++ b/src/vdaf/prio3.rs
@@ -532,7 +532,7 @@ where
     }
 }

-impl<T, P, const L: usize> Vdaf for Prio3<T, P, L>
+impl<T, P, const L: usize> Vdaf<L, 16> for Prio3<T, P, L>
 where
     T: Type,
     P: Prg<L>,
@@ -711,7 +711,7 @@ impl<F: FftFriendlyFieldElement, const L: usize> ParameterizedDecode<Prio3Prepar
     }
 }

-impl<T, P, const L: usize> Client<16> for Prio3<T, P, L>
+impl<T, P, const L: usize> Client<L, 16> for Prio3<T, P, L>
 where
     T: Type,
     P: Prg<L>,
@@ -1009,7 +1009,7 @@ where
     }
 }

-impl<T, P, const L: usize> Collector for Prio3<T, P, L>
+impl<T, P, const L: usize> Collector<L, 16> for Prio3<T, P, L>
 where
     T: Type,
     P: Prg<L>,
@@ -1509,13 +1509,21 @@ mod tests {
     #[test]
     fn roundtrip_output_share() {
         let vdaf = Prio3::new_aes128_count(2).unwrap();
-        fieldvec_roundtrip_test::<Field64, Prio3Aes128Count, OutputShare<Field64>>(&vdaf, &(), 1);
+        fieldvec_roundtrip_test::<Field64, Prio3Aes128Count, OutputShare<Field64>, 16, 16>(
+            &vdaf,
+            &(),
+            1,
+        );

         let vdaf = Prio3::new_aes128_sum(2, 17).unwrap();
-        fieldvec_roundtrip_test::<Field128, Prio3Aes128Sum, OutputShare<Field128>>(&vdaf, &(), 1);
+        fieldvec_roundtrip_test::<Field128, Prio3Aes128Sum, OutputShare<Field128>, 16, 16>(
+            &vdaf,
+            &(),
+            1,
+        );

         let vdaf = Prio3::new_aes128_histogram(2, &[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]).unwrap();
-        fieldvec_roundtrip_test::<Field128, Prio3Aes128Histogram, OutputShare<Field128>>(
+        fieldvec_roundtrip_test::<Field128, Prio3Aes128Histogram, OutputShare<Field128>, 16, 16>(
             &vdaf,
             &(),
             12,
@@ -1525,21 +1533,21 @@ mod tests {
     #[test]
     fn roundtrip_aggregate_share() {
         let vdaf = Prio3::new_aes128_count(2).unwrap();
-        fieldvec_roundtrip_test::<Field64, Prio3Aes128Count, AggregateShare<Field64>>(
+        fieldvec_roundtrip_test::<Field64, Prio3Aes128Count, AggregateShare<Field64>, 16, 16>(
             &vdaf,
             &(),
             1,
         );

         let vdaf = Prio3::new_aes128_sum(2, 17).unwrap();
-        fieldvec_roundtrip_test::<Field128, Prio3Aes128Sum, AggregateShare<Field128>>(
+        fieldvec_roundtrip_test::<Field128, Prio3Aes128Sum, AggregateShare<Field128>, 16, 16>(
             &vdaf,
             &(),
             1,
         );

         let vdaf = Prio3::new_aes128_histogram(2, &[1, 2, 3, 4, 5, 6, 7, 8, 9, 10, 11]).unwrap();
-        fieldvec_roundtrip_test::<Field128, Prio3Aes128Histogram, AggregateShare<Field128>>(
+        fieldvec_roundtrip_test::<Field128, Prio3Aes128Histogram, AggregateShare<Field128>, 16, 16>(
             &vdaf,
             &(),
             12,
```